### PR TITLE
/discussions MVP using nostr, needs correct event (and refactor to react)

### PR DESCRIPTION
**Pull request by PeerRich**
I am using the script snippet from nostrocket.org to display the comments of a nostr thread

this is just a prototype

- [ ] once @cypherhoodlum has the nostr account ready and posted a generic "Discussion Thread" event, we can change it. 
- [ ] refactor to react (get rid of Script import)
- [ ] add functionality to reply within NostrGit and not link to snort.social

CleanShot 2023-03-12 at 16 51 01@2x

### CODE REVIEW

File: src/app/[entity]/[repo]/discussions/page.tsx [hunk 1]
```diff
@@ -1,7 +1,33 @@
...
+import Script from 'next/script'
+
 export default function RepoDiscussionsPage() {
   return (
     <>
-      <h1>Discussions</h1>
+      <Script src="https://nostrocket.org/verifyBitcoinSignedMessage.js" />
+      <Script src="https://unpkg.com/nostr-tools/lib/nostr.bundle.js" />
+      <Script src="https://nostrocket.org/nostr.js" />
+      <Script src="https://nostrocket.org/comments.js" />
```
Comment (PeerRich) [inline src/app/[entity]/[repo]/discussions/page.tsx:12]: we have to eventually turn this into proper react components

there is a lot of JS fluff in there which can be simplified

File: src/app/[entity]/[repo]/discussions/page.tsx [hunk 2]
```diff
@@ -1,7 +1,33 @@
...
+      <link rel="stylesheet" href="https://nostrocket.org/comments.css" />
+      <Script
+        id="comments-init"
+        strategy="lazyOnload"
+        dangerouslySetInnerHTML={{
+          __html: `let relay = comments_init("503941a9939a4337d9aef7b92323c353441cb5ebe79f13fed77aeac615116354")`,
+        }}
```
Comment (PeerRich) [inline src/app/[entity]/[repo]/discussions/page.tsx:19]: replace this with the correct event @cypherhoodlum

File: src/app/[entity]/[repo]/discussions/page.tsx [hunk 3]
```diff
@@ -1,7 +1,33 @@
...
+      />
+      <div id="comments" className="mt-4 max-w-5xl">loading comments...</div>
+      <div className="border border-gray rounded-md w-full max-w-5xl">
+        <div className="flex text-sm bg-[#171B21] rounded-md rounded-bl-none rounded-br-none border-gray h-11 border-b pt-2 pl-2">
+          <div className="relative -bottom-px border-t border-l border-r bg-[#0F1116] rounded-md rounded-bl-none rounded-br-none border-gray h-full max-w-min flex items-center p-4">Write</div>
+          <div className="relative -bottom-px text-slate-400 h-full max-w-min flex items-center p-4">Preview</div>
+        </div>
+        <a href="https://snort.social/e/note12qu5r2vnnfpn0kdw77ujxg7r2dzped0tu7038lkh0t4vv9g3vd2qjxr9c7" target="_blank" className="p-2 block">
+          <textarea rows={4} className="w-full bg-black border border-gray rounded-md min-h-full flex h-full" />
```
Comment (PeerRich) [inline src/app/[entity]/[repo]/discussions/page.tsx:28]: since we dont have "comment" functionality yet, I am simply linking to snort.social in the meantime until we have it

File: src/app/[entity]/[repo]/discussions/page.tsx (hunk 1 above)
Comment (pierresisson) [reply src/app/[entity]/[repo]/discussions/page.tsx:12]: yeah i agree, maybe in this PR?

**Review by pierresisson — APPROVED** [review]
LGTM

**Comment by cypherhoodlum** [comment]
> yeah i agree, maybe in this PR?

Suggestion: Let's not merge this just yet (also because we may get the vanity npub still today).

**Comment by PeerRich** [comment]
feel free to add the event and merge afterwards

File: src/app/[entity]/[repo]/discussions/page.tsx (hunk 1 above)
Comment (PeerRich) [reply src/app/[entity]/[repo]/discussions/page.tsx:12]: not sure, feel free to follow up

**Comment by cypherhoodlum** [comment]
No vanity npub yet, maybe tomorrow :crossed_fingers:

**Comment by PeerRich** [comment]
> No vanity npub yet, maybe tomorrow 🤞

should we move forward with the existing nostr account?

**Comment by brycedev** [comment]
👀

**Comment by cypherhoodlum** [comment]
> should we move forward with the existing nostr account?

I'd give the miner til the end of today, but if you don't want to wait I cannot blame you :smile:

**Comment by PeerRich** [comment]
note10488mc62vzcf7yw8x0pavewdh5vmm5j4r279l9x43rss8tjzdc0s2lsrfp

File: src/app/[entity]/[repo]/discussions/page.tsx (hunk 1 above)
Comment (PeerRich) [reply src/app/[entity]/[repo]/discussions/page.tsx:12]: > yeah i agree, maybe in this PR?

I have a lot of issues with the imported scripts. maybe we should consider just making it properly with react and the nostr functions we have already

Comment (PeerRich) [reply src/app/[entity]/[repo]/discussions/page.tsx:12]: I thought I could get a quick win but it seems to cause more issues getting this to work than rebuilding it from scratch.

here is the thread: 

note10488mc62vzcf7yw8x0pavewdh5vmm5j4r279l9x43rss8tjzdc0s2lsrfp

https://iris.to/note10488mc62vzcf7yw8x0pavewdh5vmm5j4r279l9x43rss8tjzdc0s2lsrfp


let's try to display it in discussions and move all of our discussions there

Comment (PeerRich) [reply src/app/[entity]/[repo]/discussions/page.tsx:12]: @pierresisson and @e-roy wanna help refactor this PR into proper react components and use

`nostr-tools` instead?

Comment (PeerRich) [reply src/app/[entity]/[repo]/discussions/page.tsx:11]: we can get a lot of inspiration from here: https://nostrocket.org/comments.js -- should be easy to turn into a react component for us

Comment (e-roy) [reply src/app/[entity]/[repo]/discussions/page.tsx:11]: > @pierresisson and @e-roy wanna help refactor this PR into proper react components and use
> 
> `nostr-tools` instead?

I don't mind building out the UI, I just don't know how to get the data

**Comment by e-roy** [comment]
@PeerRich I think I've got a handle on the data for this.  This is what I'm seeing:

 - I can query against the `NostrGit` `npub` to get a list of discussions in this `[entity]/[repo]/discussions` page

 - each discussion can be a link to a `[entity]/[repo]/discussions/[note]` page where the discussion thread you attempted to build out above.

Problems: 
this only works for discussions initiated by the `NostrGit` account.  I'm not sure yet how to have other users initiate a discussion and to retrieve that list.

Going this route I could get a basic UI with threads set up to iterate on after.

Questions:
@cypherhoodlum what do you mean by a vanity account and what would that solve for this?  when I query the NostrGit's `npub` I get back this result

`
GitHub: "https://github.com/NostrGit/NostrGit"
about: "Manage your git repositories over Nostr."
banner: "https://nostr.build/i/nostr.build_21c55771e389acf6cb7ff1f04165373195e7d2cfa51cef08d1f708c64c4ce72a.png"
display_name: "NostrGit"
lud16: "https://getalby.com/p/cypherhoodlum"
name: "nostrgit"
nip05: "nostrgit@nostrgit.com"
nip05valid: false
picture: "https://avatars.githubusercontent.com/u/126956919?s=200&v=4"
website: "https://nostrgit.com"
`

**Comment by e-roy** [comment]
> @e-roy @cypherhoodlum fair point. a general question, how will we raise issues and discussions? (I was thinking discussions could just be a long chat, like a public telegram chat but it would also be smart to have separate discussions maybe.)
> 
> its a general question how we plan to "raise issues"

This is a great question.  In github, all 3 pages, PR, issues and discussions have a similar general structure, a main page with a list and a `[id]` page with a thread like discussion.  

With nostr protocol there are 2 `kind` that seem the most popular (I'm still trying to wrap my head around all of this, so there might be more possibilities)   

`kind 1` for Short Text Note which is the format used is this PR `note`
`kind 30023` for Long-form Content, which seems to be used for blog posts `naddr`

I'm still learning how these both are structured, so I'm not sure which is best to use. 

From what I understand there is also the possibility to create our own `NIP`????  I'm still learning more about this, but if we could, then there could be more flexibility

**Comment by cypherhoodlum** [comment]
> do we keep the same format as github does for all 3 pages, PR, issues and discussions?

I think PRs may have to be a bit different from the rest, and the focus for now should be on Issues & Discussions. PRs are so intimately git related (backend stuff). The `kind 30023` notes seem a bit overkill to me. I would stick with `kind 1` notes. Issues, and especially Discussion notes should be much more concise than a blog post. One Discussion thread may get convoluted very fast, but I like the idea of having one general Discussion thread from which more formalised Issues emerge.
In summary, I think we could try to have one general Discussion thread. Issues would be more specifically related to a bug, feature, etc.

**Comment by cypherhoodlum** [comment]
My client gets stuck at this screen for some reason (we could have a spinning animation instead of the text indicator):
Screenshot from 2023-03-17 14-45-59

**Comment by cypherhoodlum** [comment]
As @PeerRich said previously, maybe it's better to create this feature using nostr-tools without using any of the Scripts. I like the idea of a chat-like UI where users just respond to a single note. It would be a general NostrGit discussion thread that could be interacted with NostrGit itself (chat-like) or using any of the social nostr clients people already use. 

I'm thinking this PR could be closed. @PeerRich can you confirm?

File: src/app/[entity]/[repo]/discussions/page.tsx (hunk 1 above)
Comment (pqingshuang) [reply src/app/[entity]/[repo]/discussions/page.tsx:11]: Not sure this will be useful or not but I updated https://nostrocket.org/comments.js recently. I think the new js file is less messy that before in handling thread. And better to convert to js. :) sorry im just a js pleb

Comment (cypherhoodlum) [reply src/app/[entity]/[repo]/discussions/page.tsx:11]: Hi @pqingshuang, thanks for the update :+1: I'll look into it.

**Comment by PeerRich** [comment]
agree
